src/guide-rngs.md

```diff
@@ -6,8 +6,314 @@ however, often you can just use [`thread_rng`]. This function
+
+### Worst-case performance
+Because CSPRNGs usually produce a block of values into a cache, they have
+poor worst case performance (in contrast to basic PRNGs, where the
+performance is usually quite regular).
+
+## State size
+
+Simple PRNGs often use very little memory, commonly only a few words, where
+a *word* is usually either `u32` or `u64`. This is not true for all
+non-cryptographic PRNGs however, for example the historically popular
+Mersenne Twister MT19937 algorithm requires 2.5 kB of state.
+
+CSPRNGs typically require more memory; since the seed size is recommended
+to be at least 192 bits and some more may be required for the algorithm,
+256 bits would be approximately the minimum secure size. In practice,
+CSPRNGs tend to use quite a bit more, [`ChaChaRng`] is relatively small with
+136 bytes of state.
+
+## Initialization time
+
+The time required to initialize new generators varies significantly. Many
+simple PRNGs and even some cryptographic ones (including [`ChaChaRng`])
+only need to copy the seed value and some constants into their state, and
+thus can be constructed very quickly. In contrast, CSPRNGs with large state
+require an expensive key-expansion.
+
+# Quality
+
+Many basic PRNGs are not much more than a couple of bitwise and arithmetic
+operations. Their simplicity gives good performance, but also means there
+are small regularities hidden in the generated random number stream.
+
+How much do those hidden regularities matter? That is hard to say, and
+depends on how the RNG gets used. If there happen to be correlations between
+the random numbers and the algorithm they are used in, the results can be
+wrong or misleading.
+
+A random number generator can be considered good if it gives the correct
+results in as many applications as possible. The quality of PRNG
+algorithms can be evaluated to some extend analytically, to determine the
+cycle length and to rule out some correlations. Then there are empirical
+test suites designed to test how well a PRNG performs on a wide range of
+possible uses, the latest and most complete of which are [TestU01] and
+[PractRand].
+
+CSPRNGs tend to be more complex, and have an explicit requirement to be
+unpredictable. This implies there must be no obvious correlations between
+output values.
+
+### Quality stars:
+PRNGs with 3 stars or more should be good enough for any purpose.
+1 or 2 stars may be good enough for typical apps and games, but do not work
+well with all algorithms.
+
+## Period
+
+The *period* or *cycle length* of a PRNG is the number of values that can be
+generated after which it starts repeating the same random number stream.
+Many PRNGs have a fixed-size period, but for some only an expected average
+cycle length can be given, where the exact length depends on the seed.
+
+On today's hardware, even a fast RNG with a cycle length of *only*
+2<sup>64</sup> can be used for centuries before cycling. Yet we recommend a
```

Comment:
Parallel applications might achieve cycling in reasonable time.

Reply:
If you were skipping forward to get a unique stream, yes — but if I were using small fast RNGs in a parallel program I'd prefer to seed a bunch of independent ones, seeded off a strong (crypto?) master PRNG. Isn't this really about skipping?